databricks/koalas/tests/indexes/test_base.py

```diff
@@ -65,6 +66,46 @@ def test_index_basic(self):
             self.assert_eq(kdf.index, pdf.index)
             self.assert_eq(type(kdf.index).__name__, type(pdf.index).__name__)
 
+    def test_map(self):
+        kser = ks.Series([1, 2, 3], index=[1, 2, 3])
+
+        with self.assertRaisesRegex(
+            NotImplementedError, "Currently do not support input of ks.Series in Index.map"
+        ):
+            kser.index.map(ks.Series(["one", "two", "three"], index=[1, 2, 3]))
+
+        # Apply series
+        self.assert_eq(
+            kser.index.map(pd.Series(["one", "2", "three"], index=[1, 2, 3])),
+            ks.Index(["one", "2", "three"]),
+        )
+        self.assert_eq(
+            kser.index.map(pd.Series(["one", "2"], index=[1, 2])), ks.Index(["one", "2", "3"]),
+        )
+
+        # Apply dict
+        self.assert_eq(
+            kser.index.map({1: "one", 2: "two", 3: "three"}), ks.Index(["one", "two", "three"])
+        )
+        self.assert_eq(kser.index.map({1: "one", 2: "two"}), ks.Index(["one", "two", "3"]))
+        self.assert_eq(kser.index.map({1: 10, 2: 20}, return_type=int), ks.Index([10, 20, 3]))
+
+        # Apply lambda
+        self.assert_eq(kser.index.map(lambda id: id + 1, return_type=int), ks.Index([2, 3, 4]))
+        self.assert_eq(
+            kser.index.map(lambda id: id + 1.1, return_type=float), ks.Index([2.1, 3.1, 4.1])
+        )
+        self.assert_eq(
+            kser.index.map(lambda id: "{id} + 1".format(id=id), str),
+            ks.Index(["1 + 1", "2 + 1", "3 + 1"]),
+        )
+
+        kser = ks.Series([1, 2, 3, 4], index=pd.date_range("2018-04-09", periods=4, freq="2D"))
+        self.assert_eq(
+            kser.index.map(lambda id: id + DateOffset(days=1), return_type=datetime),
+            ks.Series([1, 2, 3, 4], index=pd.date_range("2018-04-10", periods=4, freq="2D")).index,
+        )
```

Comment:
Could you also add tests with `CategoricalIndex`?
You can put the tests in `test_categorical.py`.

Reply:
@ueshin I am looking at adding tests to CategoricalIndex but have some unexpected behaviour. I wonder whether you can help explain.

The current implementation for Index.map leverages `_with_new_scol` to avoid collect anything to the driver. When you have a `CategoricalIndex` such as `ks.CategoricalIndex(["a", "b", "c"])` the returned `spark_frame` is 

```
+-----------------+-----------------+
|__index_level_0__|__natural_order__|
+-----------------+-----------------+
|                0|                0|
|                1|       8589934592|
|                2|      17179869184|
+-----------------+-----------------+
```

I was expecting

```
+-----------------+-----------------+
|__index_level_0__|__natural_order__|
+-----------------+-----------------+
|                a|                0|
|                b|       8589934592|
|                c|      17179869184|
+-----------------+-----------------+
```
Is my expectation incorrect?

This seems to be caused by https://github.com/databricks/koalas/blob/master/databricks/koalas/frame.py#L510

If you have a `pdf = pd.DataFrame(index=pd.CategoricalIndex(["a", "b", "c"]))` `InternalFrame.from_pandas(pdf).spark_frame` returns  

```
+-----------------+-----------------+
|__index_level_0__|__natural_order__|
+-----------------+-----------------+
|                0|                0|
|                1|       8589934592|
|                2|      17179869184|
+-----------------+-----------------+
```

Where as if you have `pdf = pd.DataFrame(index=pd.Index(["a", "b", "c"])` `InternalFrame.from_pandas(pdf).spark_frame` returns

```
+-----------------+-----------------+
|__index_level_0__|__natural_order__|
+-----------------+-----------------+
|                a|                0|
|                b|       8589934592|
|                c|      17179869184|
+-----------------+-----------------+
```

Apologies if this is a silly question. Thanks in advance

Reply:
The behavior is expected since so far Koalas manages only `Categorical` 's `codes` in Spark and its `categories` is in the metadata, `index_dtypes` or `data_dtypes`.